datafusion/expr-common/src/signature.rs

```diff
@@ -138,6 +141,48 @@ pub enum TypeSignature {
     NullAry,
 }
 
+impl TypeSignature {
+    #[inline]
+    pub fn is_one_of(&self) -> bool {
+        matches!(self, TypeSignature::OneOf(_))
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialOrd)]
+pub enum TypeSignatureClass {
+    Timestamp,
```

Comment:
We may need to treat timestamp and timestamp with zone separately 🤔

Reply:
Unless there is function that treat timestamp differently based on timezone, otherwise it is simpler to treat them equivalently.

Reply:
I also feel like we should have another variant for `timestamp with time zone`. IMO, they are different types.
- `Timestamp(timeunit, None)` for `timestamp without time zone`.
- `Timestamp(timeunit, Some(TIMEZONE_WILDCARD)` for `timestamp with time zone`.

They can't interact with each other before applying some casting or coercion.

For example, 
https://github.com/apache/datafusion/blob/91670e2d152d5819b0f102268f5eeb299f0d6823/datafusion/functions/src/datetime/date_bin.rs#L59-L68

The `date_bin` function accepts two `Timestamp` arguments. However, if we try to simplify here, we may write something like
```
  TypeSignature::Coercible(vec![
      TypeSignatureClass::Interval,
      TypeSignatureClass::Timstamp,
      TypeSignatureClass::Timstamp,
  ]),
```
It means we can accept the SQL like `date_bin(INTERVAL 1 HOUR, timestamp_without_timezone_col, timestamp_with_timezone_col)` but I guess it's not correct. (no match the original signature)

If we have a class for `timestamp with time zone`, we can write
```rust
TypeSianture::one_of([
  TypeSignature::Coercible(vec![
      TypeSignatureClass::Interval,
      TypeSignatureClass::Timstamp,
      TypeSignatureClass::Timstamp,
  ]),
  TypeSignature::Coercible(vec![
      TypeSignatureClass::Interval,
      TypeSignatureClass::Timstamp_with_time_zone,
      TypeSignatureClass::Timstamp_with_time_zone,
  ]),
])
```
It's more close to the original signature.

Reply:
How about managing timestamps within the invoke function? Handling specific cases like Timestamp_with_time_zone or TIMEZONE_WILDCARD adds unnecessary complexity to the function's signature without providing much benefit. Instead, why not define a high-level Timestamp in the signature and handle the finer details elsewhere?

Reply:
Maybe we can refer to the behavior of `Postgres` or `DuckDB`. I listed the signature of the timestamp function `age` from their system catalogs for reference.

### Postgres
```sql
test=# SELECT proname, proargtypes FROM pg_proc WHERE proname = 'age';
 proname | proargtypes 
---------+-------------
 age     | 28
 age     | 1114
 age     | 1184
 age     | 1114 1114
 age     | 1184 1184
(5 rows)

test=# SELECT oid, typname FROM pg_type WHERE oid IN (28, 1184, 1114);
 oid  |   typname   
------+-------------
   28 | xid
 1114 | timestamp
 1184 | timestamptz
(3 rows)
```

### DuckDB
```sql
D SELECT function_name, parameter_types 
  FROM duckdb_functions() 
 WHERE function_name = 'age';
┌───────────────┬──────────────────────────────────────────────────────┐
│ function_name │                   parameter_types                    │
│    varchar    │                      varchar[]                       │
├───────────────┼──────────────────────────────────────────────────────┤
│ age           │ [TIMESTAMP]                                          │
│ age           │ [TIMESTAMP, TIMESTAMP]                               │
│ age           │ [TIMESTAMP WITH TIME ZONE, TIMESTAMP WITH TIME ZONE] │
│ age           │ [TIMESTAMP WITH TIME ZONE]                           │
└───────────────┴──────────────────────────────────────────────────────┘
```

If you try to input an unsupported type value, you may encounter an error like the following:  

```sql
D SELECT age('2001-01-01 18:00:00');
Binder Error: Could not choose a best candidate function for the function call "age(STRING_LITERAL)". In order to select one, please add explicit type casts.
    Candidate functions:
    age(TIMESTAMP WITH TIME ZONE) -> INTERVAL
    age(TIMESTAMP) -> INTERVAL

LINE 1: SELECT age('2001-01-01 18:00:00');
```

Both systems treat `TIMESTAMP` and `TIMESTAMP WITH TIME ZONE` as distinct types in the high level.  

The advantage of separating these types is that it allows for stricter type-checking when matching a function's signature. This reduces the likelihood of developers failing to correctly handle type checks when implementing timestamp functions.

Honestly, I'm not sure how complex it would become if we separated them 🤔 . If it requires significant effort, I'm fine with keeping the current design.

Reply:
My solution so far is that we use TypeSignature::Timestamp if we don't care about it has timezone or not. Fallback to `TypeSignature::Native()` if we need to tell the difference

Reply:
> My solution so far is that we use TypeSignature::Timestamp if we don't care about it has timezone or not. Fallback to `TypeSignature::Native()` if we need to tell the difference
>
Agreed. Handling Timestamps using `TypeSignatureClass::Native` is a good idea. I still think we should treat them as different types.